### DISCUSSION TITLE
Best way to print the time a report needed for creation into the report on the header or footer

### DISCUSSION THREAD

**Discussion by mabu999** [eclipse-birt/birt#1742] [Q&A] [answered]
Are there any internal variables that can be used for ?
Or may I get the Time the Report starts creating ?

**Comment by SteveSchafer-Innovent** [comment]
Off the top of my head I think this will work:

in initialize:

reportContext.setGlobalVariable("reportCreationDate", new Date().toString());

in text control in grid in page-header of master page:

&lt;VALUE-OF&gt;reportContext.getGlobalVariable("reportCreationDate")&lt;/VALUE-OF&gt;

**Comment by mabu999** [comment]
Thanks, works fine, but 
`<VALUE-OF format="MMM-dd-yy hh:mm:ss">reportContext.getGlobalVariable("reportCreationDate")</VALUE-OF>`
 shows the same date and time as 
`<value-of format="MMM-dd-yy hh:mm:ss">new Date()</value-of>`
I must look how to get an actual time, the pages rendering has finished and the page will be displayed

**Reply by mabu999** [reply]
I found a solution of my problem.

To get the actual time i use javascript and place the output via div and innerHTML in my footer

```
<font face=courier>start skript :
<VALUE-OF format="yyyy-MM-dd HH:mm:ss">reportContext.getGlobalVariable("reportCreationDate")</VALUE-OF><font>
<div id="zaehler">&nbsp;div</div>

<script type="text/javascript">	
function padToTwoDigits(num) {
    return num.toString().padStart(2, '0');
}

function formatDate(date) {
    let year = date.getFullYear();
    let month = padToTwoDigits(date.getMonth() + 1);  // Monate sind nullbasiert
    let day = padToTwoDigits(date.getDate());
    let hours = padToTwoDigits(date.getHours());
    let minutes = padToTwoDigits(date.getMinutes());
    let seconds = padToTwoDigits(date.getSeconds());

    return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

let aktuellesDatum = new Date();

document.getElementById("zaehler").innerHTML = "<font color=blue face=courier><b>stop skript  &nbsp;: " + formatDate(aktuellesDatum) + "&nbsp; </b></font>";
</script>

```

![image](https://github.com/eclipse-birt/birt/assets/122016135/3a4c67fc-6173-4eb5-b8b5-d3aa60db775b)

**Comment by hvbtup** [comment]
It does not make sense to show the time when the report generation has ended.
Because if you want to show this time, then this get-and-show-time must be part of the report generation.

Usually, generating a report should take only a few seconds. Thus if you show the time without seconds, it should be good enough.

If you really want to show a more precise time, than you'd need to get a time from your data source and do something like SET TRANSACTION READONLY (if you are using an Oracle DB, similar commands for other DBs) to ensure that all the data your data sources delivers is from exactly the same point in time.

**Comment by mabu999** [comment]
My goal it to see how long the report really runs.

Unfortunately we have some reports that runs some minutes, so it would be good to have a tool to stop the runtimes.

And I want to test the performance of the new birt versionen on AIX, so i may upgrade to 4.16 if it's faster than 4.13

**Comment by mabu999** [comment, marked as answer]
Now it works, with a little bit of enhanced javascript codeing i made a footnode, that shows me the runtime of the report:

```
<font face=courier id="startdate">start Report : <VALUE-OF format="yyyy-MM-dd HH:mm:ss">new Date()</VALUE-OF><font>
<div id="stopdate">&nbsp;div</div>
<div id="laufzeit">&nbsp;laufzeit</div>

<script type="text/javascript"> 

function padToTwoDigits(num) {
return num.toString().padStart(2, '0');
}

function formatDate(date) {
let year = date.getFullYear();
let month = padToTwoDigits(date.getMonth() + 1); // Monate sind nullbasiert
let day = padToTwoDigits(date.getDate());
let hours = padToTwoDigits(date.getHours());
let minutes = padToTwoDigits(date.getMinutes());
let seconds = padToTwoDigits(date.getSeconds());

return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

function timeToSeconds(time) {
    let parts = time.split(':');
    let hours = parseInt(parts[0], 10);
    let minutes = parseInt(parts[1], 10);
    let seconds = parseInt(parts[2], 10);
    
    return hours * 3600 + minutes * 60 + seconds;
}

function differenceInSeconds(time1, time2) {
    let seconds1 = timeToSeconds(time1);
    let seconds2 = timeToSeconds(time2);
    
    return Math.abs(seconds1 - seconds2);
}

let aktuellesDatum = new Date();
date_format_str = formatDate(aktuellesDatum);

document.getElementById("stopdate").innerHTML = "<font face=courier>stop Report &nbsp;: " +  date_format_str + "&nbsp; </font>";

var number=document.getElementById("startdate").innerHTML;  

// 8 stellen von rechts abschneiden
let startdate_kurz = number.slice(-8);
let stopdate_kurz = date_format_str.slice(-8);

let diffInSeconds = differenceInSeconds(stopdate_kurz, startdate_kurz);

document.getElementById("laufzeit").innerHTML = "<font color=blue face=courier><b>Laufzeit &nbsp;&nbsp;&nbsp; :&nbsp;" + diffInSeconds + "s </b></font>";

</script>
```


![image](https://github.com/eclipse-birt/birt/assets/122016135/ef7fda7d-c7c4-406b-acc7-d2940a6b4316)

**Comment by SteveSchafer-Innovent** [comment]
That's an interesting solution.  The start date is obtained in javascript executing in the report generation and the end date and comparison are done in javascript executing in the browser.    It does depend on outputting to HTML and those two operating in the same timezone.

I wonder if there's a way to measure report generation time strictly in report events.  I seem to remember trying this in the past and found it to be tricky.

**Comment by hvbtup** [comment]
BIRT "computes" a report basically top-down. What I would do is:
- Record the start time in the initialize() event.
- Add a (possibly) hidden element to the end of the body (not the masterpage). Get the current time here, subtract from it the start time, then output (or just log) the time delta.

This basically computes the time to collect the data for the report; it does not include the time to load and parse the rptdesign file and does not include all of the time needed in the emitter for formatting the output, but practically this should give you a a reasonable measure.

**Comment by mabu999** [comment]
My Solutions works for me. I wanted to know, how long a birt-report runs in total.

I made some performance tests between 4.13 and 4.16 on AIX 7.3 on Tomcat.  
I tested different reports and 4.16 is a "little bit" faster and the render engine for the charts is definitely faster, so I plan to update my production environment to 4.16 soon.
